docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,157 @@
+
+- Since kernel usage can fluctuate and is out of kubernetes control, `Allocatable` may vary over
+  time.
+
+#### System Reserved
+
+- **TODO**: Determine which resources *must* be specified (CPU+memory?)
+- **TODO**: Decide on format (parsed string vs. individual resource flags)
+
+`SystemReserved` is a **required** parameter specifying resources reserved for system components
+(4). It is provided as a command-line flag to the Kubelet at startup, and therefore cannot be
+changed during normal Kubelet operation (this could change in the [future](#future-work)).
+
+#### Reserved
+
+`Reserved` is an **optional** parameter specifying reserved resources (2). It is specified as a
+Kubelet command line flag using the same format as [system reserved](#system-reserved).
+
+## Issues
+
+### System reservation is smaller than system component usage
+
+Solution **TBD**
+
+**Option 1:** Do nothing (best effort). Let the system daemons overflow the reserved resources and
+  hope for the best. If the machine usage is less than Allocatable, there will be some room for
+  overflow and the node should continue to function. If the machine has been scheduled to capacity
+  (worst-case scenario), it may enter an unstable state.
+
+**Option 2:** Assume the worst and transition the node to the not-ready state. The kubelet could
+  attempt to reduce usage by evicting pods until usage comes under control. At that point the node
+  could be made available again.
+
+### Version discrepancy
+
+**API server / scheduler is not allocatable-resources aware:** If the Kubelet rejects a Pod but the
+  scheduler expects the Kubelet to accept it, the system could get stuck in an infinite loop
+  scheduling a Pod onto the node only to have Kubelet repeatedly reject it. To avoid this situation,
+  we will do a 2-stage rollout of `Allocatable`. In stage 1 (targeted for 1.2), `Allocatable` will
+  be reported by the Kubelet and the scheduler will be updated to use it, but Kubelet will continue
+  to do admission checks based on `Capacity` (same as today). In stage 2 of the rollout (targeted
+  for 1.3), the Kubelet will start doing admission checks based on `Allocatable`.
+
+**API server expects `Allocatable` but does not receive it:** If the kubelet is older and does not
+  provide `Allocatable` in the `NodeStatus`, the scheduler should fallback to the current behavior
+  and use `Capacity`.
+
+### 3rd party schedulers
+
+The community should be notified that an update to schedulers is recommended, but if a scheduler is
+not updated it falls under the above case of "scheduler is not allocatable-resources aware".
+
+## Future work
+
+- Convert kubelet flags to Config API - Cleaner API, better contract, etc. See
+  [#12245](https://github.com/kubernetes/kubernetes/issues/12245). Once the flags are migrated, we
+  may consider allowing updates to reservations.
+- Set cgroup limits according system reservation - as described in the [overview](#overview)
```

Comment:
Kubelet should not perform any re-organization of system daemons in certain
environments. We will have to validate that the limits provided to kubelet
are not greater than the cgroup limits imposed by the admin or user.

On Fri, Nov 13, 2015 at 2:42 PM, Tim St. Clair <notifications@github.com>
wrote:

> In docs/proposals/node-allocatable.md
> <https://github.com/kubernetes/kubernetes/pull/17201#discussion_r44842946>
> :
>
> > +
> > +**API server expects `Allocatable` but does not receive it:** If the kubelet is older and does not
> > +  provide `Allocatable` in the `NodeStatus`, the scheduler should fallback to the current behavior
> > +  and use `Capacity`.
> > +
> > +### 3rd party schedulers
> > +
> > +The community should be notified that an update to schedulers is recommended, but if a scheduler is
> > +not updated it falls under the above case of "scheduler is not allocatable-resources aware".
> > +
> > +## Future work
> > +
> > +- Convert kubelet flags to Config API - Cleaner API, better contract, etc. See
> > +  [#12245](https://github.com/kubernetes/kubernetes/issues/12245). Once the flags are migrated, we
> > +  may consider allowing updates to reservations.
> > +- Set cgroup limits according system reservation - as described in the [overview](#overview)
>
> Acknowledged. I don't think this changes anything with the rest of this
> proposal, since that's already the case today? Do you just mean that when
> we start setting limits we need to account for limits already being set?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17201/files#r44842946>.
>